react/react1/week3/time-registration/src/SubmitShifts.js

```diff
@@ -0,0 +1,96 @@
+import { useEffect, useState } from "react";
+import "./shifts.css";
+import { ShiftListItem } from "./ShiftOverview";
+
+function FancyBorder(props) {
+  return <div className="fancy-border">{props.children}</div>;
+}
+export function SubmitShifts({ fetchUrl }) {
+  const [list, setList] = useState([]);
+  const [nameInput, setNameInput] = useState("");
+  const [startInput, setStartInput] = useState("");
+  const [endInput, setEndInput] = useState("");
+  useEffect(() => {
+    fetch(fetchUrl)
+      .then((response) => response.json())
+      .then(setList);
+  }, []);
+  const addShift = () => {
+    const newId = new Date().getTime();
+    const newTodo = {
+      name: nameInput,
+      start: startInput,
+      end: endInput,
+      id: newId,
+    };
+    setList((prevTodos) => {
+      return [...prevTodos, newTodo];
+    });
+  };
+  console.log(list);
+  const display = list.map((shift) => (
+    <ShiftListItem
+      name={shift.name}
+      start={shift.start}
+      end={shift.end}
+      key={shift.newId}
+    />
+  ));
+  const searchEmployee = (e) => {
+    const value = e.target.value.toLowerCase();
+    const searchName = list.filter((shift) =>
+      shift.name.toLowerCase().includes(value)
+    );
+    setList(searchName);
+  };
+
+  const FormPart = (
+    <div>
+      <h2>Submit Shift</h2>
+      <form>
+        <label>Name</label>
+        <input
+          type="text"
+          placeholder="Name"
+          value={nameInput}
+          onChange={(e) => setNameInput(e.target.value)}
+        ></input>
+        <label>Start Time</label>
+        <input
+          type="datetime-local"
+          value={startInput}
+          onChange={(e) => setStartInput(e.target.value)}
+        ></input>
+        <label>End Time</label>
+        <input
+          type="datetime-local"
+          value={endInput}
+          onChange={(e) => setEndInput(e.target.value)}
+        ></input>
+      </form>
+      <button className="button" onClick={addShift}>
+        Save Shift
+      </button>
+    </div>
+  );
```

Comment:
Assigning FormPart to a variable inside a component works, but one of the main benefits of using React is that we can create individual components. Doing so gives us more control over our code and it makes it easier to resuse our components.

You can convert FormParts into it's own component like this:

```
const FormPart = ({ setList }) => {
  const [nameInput, setNameInput] = useState('');
  const [startInput, setStartInput] = useState('');
  const [endInput, setEndInput] = useState('');

  const addShift = () => {
    const newId = new Date().getTime();
    const newTodo = {
      name: nameInput,
      start: startInput,
      end: endInput,
      id: newId
    };
    setList((prevTodos) => {
      return [...prevTodos, newTodo];
    });
  };

  return (
    <div>
      <h2>Submit Shift</h2>
      <form>
        <label>Name</label>
        <input
          type="text"
          placeholder="Name"
          value={nameInput}
          onChange={(e) => setNameInput(e.target.value)}
        ></input>
        <label>Start Time</label>
        <input
          type="datetime-local"
          value={startInput}
          onChange={(e) => setStartInput(e.target.value)}
        ></input>
        <label>End Time</label>
        <input
          type="datetime-local"
          value={endInput}
          onChange={(e) => setEndInput(e.target.value)}
        ></input>
      </form>
      <button className="button" onClick={addShift}>
        Save Shift
      </button>
    </div>
  );
};
```

As you can see this also allows us to place all of the useState inside of that component. We can keep the state relevant to our component inside that component. The only piece of code we need is `setList` and we can pass that in using props:

```
<FancyBorder>
  <FormPart setList={setList} />
</FancyBorder>

```

You can do the same with `ShiftsPart` and `display`.

Reply:
hi Bjorn thank you for your feedback. the method that you have suggested is very efficient. but I have confusion about which props should I give to which component. and which component will have which callback function as parameter. like you suggested I have made shiftsPart an individual component and given it {setList} a parameter and then use {setList} as prop. again I am not clear about display component.